# Librerías

In [1]:
import pandas as pd

# Configuraciones

In [2]:
pd.set_option('display.max_columns', None)

# Ingesta de Datos

In [3]:
df_clin_hosp = pd.read_excel('glosas_clasificadas_me2l.xlsx')

# EDA

## Clínicas y Hospitales

In [4]:
df_clin_hosp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18087 entries, 0 to 18086
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Unnamed: 0.1            18087 non-null  int64         
 1   Unnamed: 0              18087 non-null  int64         
 2   Documento compras       18087 non-null  int64         
 3   Grupo de compras        18087 non-null  object        
 4   Grupo de liberación     16711 non-null  object        
 5   Posición                18087 non-null  int64         
 6   Fecha documento         18087 non-null  datetime64[ns]
 7   Nombre del proveedor    18087 non-null  object        
 8   Material                16798 non-null  float64       
 9   Texto breve             18086 non-null  object        
 10  Cantidad de pedido      18087 non-null  float64       
 11  Cantidad en UMA         18087 non-null  float64       
 12  Cantidad de imputación  18087 non-null  float6

In [5]:
df_clin_hosp.head()

Unnamed: 0.1  Unnamed: 0  Documento compras Grupo de compras  \
0         30801  4561463896         4561463896              S08   
1         30802  4561463896         4561463896              S08   
2         30803  4561463896         4561463896              S08   
3         30876  4561463932         4561463932              S13   
4         31003  4561463961         4561463961              S17   

  Grupo de liberación  Posición Fecha documento  \
0                  P3        10      2024-11-02   
1                  P3        20      2024-11-02   
2                  P3        30      2024-11-02   
3                  P3        10      2024-11-04   
4                  P3        10      2024-11-04   

                 Nombre del proveedor     Material  \
0        616021427  HOSPITAL DE PEUMO  115800002.0   
1        616021427  HOSPITAL DE PEUMO  115800002.0   
2        616021427  HOSPITAL DE PEUMO  115800002.0   
3   116996766  GLORIA LUZ SILVA OSSES          NaN   
4  762619059  CENTRO MEDICO LOS ANDES  115800115.0   

                                Texto breve  Cantidad de pedido  \
0                      ATENCION DE URGENCIA                 1.0   
1                      ATENCION DE URGENCIA                 1.0   
2                      ATENCION DE URGENCIA                 1.0   
3             Hon. Arsenalera- Juan Mondaca                 1.0   
4  CMUPM;HOSP;EVELYN PRISCILA VARGAS TORRES                 1.0   

   Cantidad en UMA  Cantidad de imputación Unidad medida pedido  Precio neto  \
0              1.0                     1.0                   UN        33110   
1              1.0                     1.0                   UN        54790   
2              1.0                     1.0                   UN        17210   
3              0.0                     1.0                   UN        35000   
4              1.0                     1.0                   UN       425030   

  Moneda  Centro  Ind.liberación Centro de coste  Cuenta de mayor  \
0    CLP    1700             2.0           61121           510601   
1    CLP    1700             2.0           61121           510601   
2    CLP    1700             2.0           61121           510601   
3    CLP    2200             2.0           82132           510601   
4    CLP    2600             2.0          106131           510601   

          Orden  Indicador de borrado Indicador de borrado.1  \
0  202409180966                   NaN                    NaN   
1  202409190828                   NaN                    NaN   
2  202409190231                   NaN                    NaN   
3  202409132786                   NaN                    NaN   
4  202308847774                   NaN                    NaN   

  Estado liberación  Cantidad base Cl.documento compras  Unnamed: 25  \
0                 X              1                 PSEM          NaN   
1                 X              1                 PSEM          NaN   
2                 X              1                 PSEM          NaN   
3                 X              1                 PSPF          NaN   
4               XXX              1                 PSEM          NaN   

           nombre_cuenta    Clasificacion_Phi4  
0  Clínicas y Hospitales  Atención de urgencia  
1  Clínicas y Hospitales  Atención de urgencia  
2  Clínicas y Hospitales  Atención de urgencia  
3  Clínicas y Hospitales      Ley de pacientes  
4  Clínicas y Hospitales       Hospitalización

In [6]:
df_clin_hosp['Cantidad de imputación'].value_counts()

Cantidad de imputación
1.00      17934
2.00         45
3.00         14
4.00         11
31.00         9
5.00          8
12.00         6
6.00          6
240.00        5
40.00         5
60.00         5
14.00         4
28.00         4
0.00          4
10.00         3
30.00         3
8.00          3
7.00          2
1.75          2
100.00        2
51.00         1
19.00         1
55.00         1
58.00         1
15.00         1
52.00         1
9.00          1
11.00         1
70.00         1
56.00         1
13.00         1
49.00         1
Name: count, dtype: int64

In [7]:
df_clin_hosp[df_clin_hosp['Cantidad de imputación']==31.00]

Unnamed: 0.1  Unnamed: 0  Documento compras Grupo de compras  \
1618          46277  4561466839         4561466839              S07   
1619          46278  4561466840         4561466840              S07   
1621          46314  4561466844         4561466844              S07   
9220         114300  4561479400         4561479400              S07   
9249         114443  4561479447         4561479447              S07   
9251         114445  4561479448         4561479448              S07   
13590        155060  4561487275         4561487275              S07   
13591        155098  4561487284         4561487284              S07   
13594        155188  4561487295         4561487295              S07   

      Grupo de liberación  Posición Fecha documento  \
1618                   P3        10      2024-11-14   
1619                   P3        10      2024-11-14   
1621                   P3        10      2024-11-14   
9220                   P3        10      2025-01-09   
9249                   P3        10      2025-01-09   
9251                   P3        10      2025-01-09   
13590                  P3        10      2025-02-18   
13591                  P3        10      2025-02-18   
13594                  P3        10      2025-02-18   

                     Nombre del proveedor  Material  \
1618   611080000  CAJA DE PREVISION DE LA       NaN   
1619   611080000  CAJA DE PREVISION DE LA       NaN   
1621   611080000  CAJA DE PREVISION DE LA       NaN   
9220   611080000  CAJA DE PREVISION DE LA       NaN   
9249   611080000  CAJA DE PREVISION DE LA       NaN   
9251   611080000  CAJA DE PREVISION DE LA       NaN   
13590  611080000  CAJA DE PREVISION DE LA       NaN   
13591  611080000  CAJA DE PREVISION DE LA       NaN   
13594  611080000  CAJA DE PREVISION DE LA       NaN   

                       Texto breve  Cantidad de pedido  Cantidad en UMA  \
1618    032 OCT/24 DCA 8.385.113-9                31.0              0.0   
1619   032 OCT/24 DCA 16.678.673-8                31.0              0.0   
1621    032 OCT/24 DCA 7.663.163-8                31.0              0.0   
9220   032 DIC/24 DCA 16.678.673-8                31.0              0.0   
9249    032 DIC/24 DCA 7.663.163-8                31.0              0.0   
9251    032 DIC/24 DCA 8.385.113-9                31.0              0.0   
13590   032 ENE/25 DCA 8.385.113-9                31.0              0.0   
13591   032 ENE/24 DCA 7.663.163-8                31.0              0.0   
13594  032 ENE/25 DCA 16.678.673-8                31.0              0.0   

       Cantidad de imputación Unidad medida pedido  Precio neto Moneda  \
1618                     31.0                   UN       259646    CLP   
1619                     31.0                   UN       259646    CLP   
1621                     31.0                   UN       259646    CLP   
9220                     31.0                   UN       259646    CLP   
9249                     31.0                   UN       259646    CLP   
9251                     31.0                   UN       259646    CLP   
13590                    31.0                   UN       259646    CLP   
13591                    31.0                   UN       259646    CLP   
13594                    31.0                   UN       259646    CLP   

       Centro  Ind.liberación Centro de coste  Cuenta de mayor         Orden  \
1618     1600             2.0           50166           510601  201201499029   
1619     1600             2.0           50166           510601  201301709133   
1621     1600             2.0           50166           510601  200213016195   
9220     1600             2.0           50166           510601  201301709133   
9249     1600             2.0           50166           510601  200213016195   
9251     1600             2.0           50166           510601  201201499029   
13590    1600             2.0           50166           510601  201201499029   
13591    1600             2.0           50166           510601

Debo multiplicar cantidad por precio neto

In [8]:
df_clin_hosp.Moneda.value_counts()

Moneda
CLP    18087
Name: count, dtype: int64

In [9]:
df_clin_hosp['total_clp'] = df_clin_hosp['Cantidad de imputación'] * df_clin_hosp['Precio neto']

En esta cuenta solo hay CLP no debo preocuparme por las UF

In [10]:
import plotly.express as px

# Contamos las clasificaciones
clasificaciones = df_clin_hosp.Clasificacion_Phi4.value_counts()

# Creamos el gráfico de torta interactivo
fig = px.pie(
    values=clasificaciones.values, 
    names=clasificaciones.index, 
    title='Distribución de Clasificaciones Phi4',
    hole=0.3
)

# Personalizamos el diseño del gráfico
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.update_layout(showlegend=True)

# Mostramos el gráfico
fig.show()

In [11]:
# Contamos el número de ocurrencias por cada clasificación Phi4
n_por_clasificacion = df_clin_hosp['Clasificacion_Phi4'].value_counts()

# Ordenamos de mayor a menor
n_por_clasificacion = n_por_clasificacion.sort_values(ascending=False)

# Calculamos el porcentaje de cada clasificación
porcentaje_por_clasificacion = (n_por_clasificacion / n_por_clasificacion.sum()) * 100

# Creamos una tabla con el N y el % por categoría
tabla_n = n_por_clasificacion.reset_index()
tabla_n.columns = ['Clasificación Phi4', 'N']
tabla_n['%'] = porcentaje_por_clasificacion.values

# Formateamos los números para que tengan separador de miles y el porcentaje con dos decimales
tabla_n['N'] = tabla_n['N'].apply(lambda x: f"{x:,}")
tabla_n['%'] = tabla_n['%'].apply(lambda x: f"{x:.2f}%")

# Mostramos la tabla
print(tabla_n)


           Clasificación Phi4      N       %
0        Atención de urgencia  8,718  48.20%
1  Profesionales y honorarios  2,468  13.65%
2              Sin clasificar  1,873  10.36%
3  Diagnóstico / imagenología  1,598   8.84%
4      Medicamentos e insumos  1,595   8.82%
5             Hospitalización  1,565   8.65%
6  Procedimientos quirúrgicos    166   0.92%
7            Ley de pacientes    104   0.57%


In [12]:
# Calculamos los montos totales en CLP por cada clasificación Phi4
montos_por_clasificacion = df_clin_hosp.groupby('Clasificacion_Phi4')['total_clp'].sum()

# Creamos el gráfico de torta interactivo para los montos totales
fig_montos = px.pie(
    values=montos_por_clasificacion.values, 
    names=montos_por_clasificacion.index, 
    title='Distribución de Montos Totales CLP por Clasificación Phi4',
    hole=0.3
)

# Personalizamos el diseño del gráfico
fig_montos.update_traces(textposition='inside', textinfo='percent+label')
fig_montos.update_layout(showlegend=True)

# Mostramos el gráfico
fig_montos.show()


In [15]:

# Calculamos los montos totales en CLP por cada clasificación Phi4
montos_por_clasificacion = df_clin_hosp.groupby('Clasificacion_Phi4')['total_clp'].sum()

# Ordenamos los montos de mayor a menor
montos_por_clasificacion = montos_por_clasificacion.sort_values(ascending=False)

# Calculamos el porcentaje de cada monto respecto al total
porcentaje_montos = (montos_por_clasificacion / montos_por_clasificacion.sum()) * 100

# Creamos una tabla con los montos y el porcentaje por categoría
tabla_montos = montos_por_clasificacion.reset_index()
tabla_montos.columns = ['Clasificación Phi4', 'Monto Total CLP']
tabla_montos['% del total del monto gastado'] = porcentaje_montos.values

# Formateamos los montos para que no usen notación científica y tengan separador de miles
tabla_montos['Monto Total CLP'] = tabla_montos['Monto Total CLP'].apply(lambda x: f"{x:,.0f}")
tabla_montos['% del total del monto gastado'] = tabla_montos['% del total del monto gastado'].apply(lambda x: f"{x:.2f}%")

# Mostramos la tabla
print(tabla_montos)


           Clasificación Phi4 Monto Total CLP % del total del monto gastado
0             Hospitalización   2,117,395,474                        37.02%
1        Atención de urgencia   1,004,106,079                        17.56%
2  Profesionales y honorarios     791,467,726                        13.84%
3      Medicamentos e insumos     747,361,157                        13.07%
4              Sin clasificar     492,246,904                         8.61%
5  Diagnóstico / imagenología     448,466,107                         7.84%
6  Procedimientos quirúrgicos      91,828,848                         1.61%
7            Ley de pacientes      26,181,300                         0.46%


Del total de glosas (18.087) hay un 10,4% (1.873) que no se pueden clasificar debido a falta de información contextual. Este 10,4% de las glosas se traduce en montos de $492.246.904.

La categoría que posee más etiquetas es 'Atención de urgencia' (Representa un 48,2% de todas las observaciones), y corresponden a un monto de $1.004.106.079.

Mientras que 'Hospitalización' aunque representa un modesto 8,7% de las observaciones, estas representan la categoría en que más se gastó, pues los montos ascienden a $2.117.395.474 (un 37% del gasto total)

Se incorporaron tecnologías de inteligencia artificial (concretamente, un LLM llamado 'phi4' de Microsoft) para procesar y clasificar las glosas de texto de ciertas cuentas de gasto. 

Se definió el objetivo de aprovechar la información disponible en las glosas de texto de las cuentas de gasto (de la ME2L (Procedimiento en SAP)) y para esto, lo que se hizo fue definir un grupo de cuentas en base a los montos que suponian, y luego para cada una se hizo un análisis de texto (se analizaron nubes de palabras con unigramas y brigramas y se le entregaron a un modelo de lenguaje (GPT-4o)) para que determinara categorias posibles de clasificar en base a los textos disponibles.

En base a este análisis, se definieron potenciales categorías de clasificación para cada cuenta de gasto, para ser validadas junto al equipo de control financiero.

Se comenzó por la cuenta de Clínincas y Hospitales, las categorías que se definieron en concreto fueron:

- Atención de urgencia  
- Hospitalización       
- Diagnóstico / imagenología
- Procedimientos quirúrgicos 
- Medicamentos e insumos 
- Profesionales y honorarios 
- Ley de pacientes     

Para identificar estas categogías en el texto, se incorporaron tecnologías de inteligencia artificial (concretamente, un LLM llamado 'phi4' de Microsoft) para procesar e identificar en el texto patrones o indicios a estas categorías y hacer una clasificación automatizada del texto.

Se definió una instrucción que sirvió como prompt de sistema, del tipo: 

> Eres un asistente experto en clasificar cuentas de gastos médicos, tu objetivo es determinar, en base al texto proporcionado, a cual de las siguientes categorías corresponde:
>> - Atención de urgencia
>> - Hospitalización
>> - Diagnóstico / imagenología
>> - Procedimientos quirúrgicos
>> - Medicamentos e insumos
>> - Profesionales y honorarios
>> - Ley de pacientes
>>> Considera el contexto completo del texto para realizar la clasificación. Responde únicamente con el nombre de la categoría seleccionada de la lista anterior, sin añadir explicaciones, introducciones o texto adicional. Si el texto no corresponde claramente a ninguna de estas categorías, responde 'Sin clasificar'.

Cada glosa de texto de procesó de forma independiente, con el modelo utilizando estas instrucciones.

Los resultados obtenidos fueron los siguientes:

Número de Casos (órdenes de compra)
| Clasificación Phi4           |     N   |    %   |
|-----------------------------|---------|--------|
| Atención de urgencia        |  8.718  | 48.20% |
| Profesionales y honorarios  |  2.468  | 13.65% |
| Sin clasificar              |  1.873  | 10.36% |
| Diagnóstico / imagenología  |  1.598  |  8.84% |
| Medicamentos e insumos      |  1.595  |  8.82% |
| Hospitalización             |  1.565  |  8.65% |
| Procedimientos quirúrgicos  |    166  |  0.92% |
| Ley de pacientes            |    104  |  0.57% |


Montos por categoría:
| Clasificación Phi4           | Monto Total CLP | % del total del monto gastado |
|-----------------------------|------------------|-------------------------------|
| Hospitalización             | 2.117.395.474    | 37.02%                        |
| Atención de urgencia        | 1.004.106.079    | 17.56%                        |
| Profesionales y honorarios  |   791.467.726    | 13.84%                        |
| Medicamentos e insumos      |   747.361.157    | 13.07%                        |
| Sin clasificar              |   492.246.904    |  8.61%                        |
| Diagnóstico / imagenología  |   448.466.107    |  7.84%                        |
| Procedimientos quirúrgicos  |    91.828.848    |  1.61%                        |
| Ley de pacientes            |    26.181.300    |  0.46%                        |


Como se puede observar: Del total de glosas (18.087) hay un 10,4% (1.873) que no se pueden clasificar debido a falta de información contextual. Este 10,4% de las glosas se traduce en montos de $492.246.904.

La categoría más frecuente en términos de n de casos es 'Atención de urgencia' (Representa un 48,2% de todas las observaciones), y corresponden a un monto de $1.004.106.079.

Mientras que 'Hospitalización' aunque representa un modesto 8,7% de las observaciones, estas representan la categoría en que más se gastó, pues los montos ascienden a $2.117.395.474 (un 37% del gasto total)

